src/toolchain.ml

```diff
@@ -0,0 +1,161 @@
+open Import
+
+type t =
+  { project_sandbox : Sandbox.t option
+  ; opam : Opam.t * Opam.Switch.t
+  }
+
+let get_command ?(args = []) { opam = opam, switch; project_sandbox } bin =
+  match project_sandbox with
+  | Some sandbox ->
+    let open Promise.Syntax in
+    let+ has_command = Sandbox.has_command sandbox bin in
+    if has_command then
+      Sandbox.get_command sandbox bin args
+    else
+      Opam.exec opam ~switch ~args:(bin :: args)
+  | _ -> Promise.return @@ Opam.exec opam ~switch ~args:(bin :: args)
+
+module Tool = struct
+  let lsp_command ?args toolchain = get_command toolchain "ocamllsp" ?args
+
+  let dune_command toolchain args = get_command toolchain "dune" ~args
+
+  let ocamlformat_command ?args toolchain =
+    get_command toolchain "ocamlformat" ?args
+
+  let ocamlc_command ?args toolchain = get_command toolchain "ocamlc" ?args
+
+  let all toolchain =
+    [ ("ocaml-base-compiler", ocamlc_command toolchain ~args:[ "--version" ])
+    ; ("ocaml-lsp-server", lsp_command toolchain ~args:[ "--version" ])
+    ; ("ocamlformat", ocamlformat_command toolchain ~args:[ "--version" ])
+    ]
+
+  let missing_command_opt binary cmd =
+    let open Promise.Syntax in
+    let+ output =
+      let open Promise.Result.Syntax in
+      let* command = Cmd.check cmd in
+      Cmd.output command
+    in
+    match output with
+    | Ok _ -> None
+    | Error _ -> Some binary
+
+  let detect_missing toolchain =
+    all toolchain
+    |> List.map ~f:(fun (binary, cmd) ->
+           let open Promise.Syntax in
+           let* cmd = cmd in
+           missing_command_opt binary cmd)
+    |> Promise.all_list
+    |> Promise.map List.filter_opt
+end
+
+let compiler_of_version v =
+  match (Platform.t, Platform.arch) with
+  | Win32, X32 -> "ocaml-base-compiler." ^ v ^ "+mingw32c"
+  | Win32, _ -> "ocaml-base-compiler." ^ v ^ "+mingw64c"
```

Comment:
@mnxn Do you know if there's going to be any issue using a 64 bits version of the compiler in a 32 bits Cygwin?

Reply:
Before we worry about this, let's make sure there are even 32 bit users out there. I highly doubt there are any left.

Reply:
Based on https://fdopen.github.io/opam-repository-mingw/installation/, it says "you can use [either archive] with either the 32-bit or 64-bit version of cygwin". The ocaml compiler isn't linked against cygwin1.dll either, so the architecture of cygwin shouldn't matter.

What does matter is the architecture of the system, which you handle how I expect here.

The only thing that might need changing is the prefix. On the same [installation page](https://fdopen.github.io/opam-repository-mingw/installation/), the documentation refers to `ocaml-variants` instead of `ocaml-base-compiler`.